AbstractSQLOptimiser.ts

```diff
@@ -0,0 +1,1217 @@
+import * as _ from 'lodash';
+
+import { AbstractSqlQuery } from './AbstractSQLCompiler';
+import { Dictionary } from 'lodash';
+
+type MatchFn = (args: AbstractSqlQuery) => AbstractSqlQuery;
+
+const escapeForLike = (str: AbstractSqlQuery) => [
+	'Replace',
+	[
+		'Replace',
+		['Replace', str, ['EmbeddedText', '\\'], ['EmbeddedText', '\\\\']],
+		['EmbeddedText', '_'],
+		['EmbeddedText', '\\_'],
+	],
+	['EmbeddedText', '%'],
+	['EmbeddedText', '\\%'],
+];
+
+const comparisons = {
+	Equals: ' = ',
+	GreaterThan: ' > ',
+	GreaterThanOrEqual: ' >= ',
+	LessThan: ' < ',
+	LessThanOrEqual: ' <= ',
+	NotEquals: ' != ',
+	Like: ' LIKE ',
+};
+
+const CheckErr = (e: Error) => {
+	if (e instanceof SyntaxError || e instanceof TypeError) {
+		throw e;
+	}
+};
+
+type FnArgs<T> = T extends (...args: infer U) => any ? U : any[];
+
+const Helper = <F extends (...args: any[]) => any>(fn: F) => {
+	return (...args: FnArgs<F>): ReturnType<F> => {
+		const result = fn(...args);
+		helped = true;
+		return result;
+	};
+};
+
+type RewriteFn = (args: AbstractSqlQuery[]) => AbstractSqlQuery;
+
+const rewriteMatch = (
+	name: string,
+	matchers: MatchFn[],
+	rewriteFn: RewriteFn,
+): MatchFn => args => {
+	checkArgs(name, args, matchers.length);
+	return rewriteFn(
+		args.map((arg, index) => {
+			// if (!_.isArray(arg)) {
+			// 	throw new SyntaxError(`Expected AbstractSqlQuery array but got ${typeof arg}`)
+			// }
+			// Hack because of _.identity cases
+			return matchers[index](arg as AbstractSqlQuery);
+		}),
+	);
+};
+
+const matchArgs = (name: string, ...matchers: MatchFn[]): MatchFn =>
+	rewriteMatch(name, matchers, args => [name, ...args]);
+
+const tryMatches = (...matchers: MatchFn[]): MatchFn => {
+	return args => {
+		let err;
+		for (const matcher of matchers) {
+			try {
+				return matcher(args);
+			} catch (e) {
+				err = e;
+			}
+		}
+		throw err;
+	};
+};
+
+const AnyValue: MatchFn = args => {
+	const [type, ...rest] = args;
+	if (type === 'Case') {
+		return typeRules[type](rest);
+	}
+
+	try {
+		return TextValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
+	try {
+		return NumericValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
+	try {
+		return BooleanValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
+	try {
+		return DateValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
+	try {
+		return JSONValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
+	try {
+		return DurationValue(args);
+	} catch (e) {
+		CheckErr(e);
+	}
```

Comment:
This could be written with a `for valueType of [TextValue, NumericValue, ...] {`